### PR TITLE
fea(#10511): add --post-renderer-args to support assign args to PostR…

### LINKED ISSUE #10511: can  `--post-renderer`  support  assign  the cmd args into  `executable` ?
now helm  `--post-renderer` only accept  the `executable` in `PATH`;  

but it is very  necessary that  the `post-renderer-executable`   worked with some args ,for example change the docker.io registry into private docker registry when deploy the chart without internet access.

can helm can add `args` to assign the args in to  `post-renderer-executable` such as ` --post-renderer-args` 
```
helm install mychart stable/wordpress --post-renderer ./path/to/render_executable --post-renderer-args "--args_to_render_executable  "
````
if this proposal be ok, i d like  to contribute a PR.

### PR DESCRIPTION
**Pull request by tanguofu**
…enderer exec

closes #10511 

**What this PR does / why we need it**:

Now helm  `--post-renderer` only accept  the `executable` in `PATH`;  

but it is very  necessary that  the `post-renderer-executable`   worked with some args ,for example change the docker.io registry into private docker registry when deploy the chart without internet access.

can helm can add `args` to assign the args in to  `post-renderer-executable` such as ` --post-renderer-args` 
```
helm install mychart stable/wordpress --post-renderer ./path/to/render_executable --post-renderer-args "--args_1_to_render_executable" --post-renderer-args "--args_2_to_render_executable"
````

**Special notes for your reviewer**:
this implement to make the less change of the helm code.  it is  more compatibility than  #9047.

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

File: pkg/postrender/exec.go
Comment (marckhouzam) [inline pkg/postrender/exec.go:37]: This is a public API so we cannot change the signature of the function.
You can either create a new function that will satisfy your requirement, or you will need to use the current one without changing its signature.

We need to do that to avoid breaking other programs using helm's API.

**Review by marckhouzam — CHANGES_REQUESTED** [review]
Thanks for the PR.  I've haven't looked at it all but let's first address the requirement of backwards-compatibility.  Please see the comment within the review.

File: pkg/postrender/exec.go
Comment (tanguofu) [reply pkg/postrender/exec.go]: i got. i had add a new function `NewExecWithArgs` and keep back the `func NewExec(binaryPath string) (PostRenderer, error) `
thanks.

File: cmd/helm/flags.go
Comment (marckhouzam) [inline cmd/helm/flags.go:162]: Since this flag can be repeated, it would be nice to let the shell completion logic know that it can suggest the flag more than once.  To achieve this you must include the string "Array" or "Slice" in the `Type()`.  You can see where the Cobra underlying library checks this: https://github.com/spf13/cobra/blob/cb9d7b1cec87c2bb005c6e2790553bcd629bc542/completions.go#L329

So can you `return "postrenderer-argsSlice"` here?

If you want to test the difference you can build your new helm and do the following:
```
$ helm __complete install --post-renderer echo --post-renderer-args arg1 --p
```
This command returns the possible completions for the specified input.  You will want to see that the `--post-renderer-args` keeps being suggested as a completion (but not the `--post-renderer` flag)

File: pkg/postrender/exec.go
Comment (marckhouzam) [inline pkg/postrender/exec.go:45]: I realized there is a better way to be backwards-compatible.  Instead of adding a new function `NewExecWithArgs()`, you can instead change the existing one, from `func NewExec(binaryPath string) (PostRenderer, error)` to `func NewExec(binaryPath string, args ...string) (PostRenderer, error)`.

This approach allows to call `NewExec(path, args...)` to specify arguments, but also allows calling `NewExec(path)` like before.

Sorry about not having thought about it before.

**Review by marckhouzam — CHANGES_REQUESTED** [review]
Nice job with this.  A couple of comments and we should be good.

File: pkg/postrender/exec.go
Comment (tanguofu) [reply pkg/postrender/exec.go:45]: i got it,  this is best way! i did not found yet, very thanks for u suggestion, i will fix those later.

**Comment by tanguofu** [comment]
@marckhouzam  i am  sorry for update take a few days,  may you review again? thanks  for your instruction.

File: pkg/postrender/exec_test.go
Comment (marckhouzam) [inline pkg/postrender/exec_test.go:140]: Can you call the argument `arg` instead?  I find reusing `FOOTEST` to be confusing.

File: pkg/postrender/exec_test.go
Comment (marckhouzam) [inline pkg/postrender/exec_test.go:157]: Can you call the arguments `arg1` and `arg2` instead?  I find reusing `FOOTEST` to be confusing.

File: cmd/helm/flags.go
Comment (marckhouzam) [inline cmd/helm/flags.go:120]: Can you change this string to be:
"an argument to the post-renderer  (can specify multiple)"

File: cmd/helm/flags.go
Comment (marckhouzam) [inline cmd/helm/flags.go:167]: This is a bit of a corner case, but you had it better in your previous version.
Even when `p.options.binaryPath == ""` we should still call `p.options.args = append(p.options.args, val)`.  This allows to use the `--post-renderer-args` before the `--post-renderer` flag on the command-line.
For example, if I run:
```
$ bin/helm template test bitnami/nginx --post-renderer-args arg1  --post-renderer echo --post-renderer-args arg2
arg2
```
Notice that arg1 is ignored because it was specified before `--post-renderer`

**Review by marckhouzam — CHANGES_REQUESTED** [review]
Nicely done.  One corner case to fix and I think we are done.

File: cmd/helm/flags.go
Comment (tanguofu) [reply cmd/helm/flags.go:167]: got! .  i will fix all of these.

File: cmd/helm/flags.go
Comment (marckhouzam) [inline cmd/helm/flags.go:170]: I think that for the arguments to the post-renderer, an empty argument could be valid.
I don't have an example, but a post-renderer defined by a user may accept empty arguments specified like this:
```
$ bin/helm template test bitnami/nginx --post-renderer myrenderer --post-renderer-args ''

```
Notice that is this different than forgetting to add an argument, in which case an error will be shown:
```
$ bin/helm template test bitnami/nginx --post-renderer echo --post-renderer-args
Error: flag needs an argument: --post-renderer-args
```
So, what do you think of removing the check for `val == ""` here?

Comment (tanguofu) [reply cmd/helm/flags.go:167]: i got !
very thanks your  detailed and comprehensive advice, that help me  learn much more code skills.
many thanks!  and `helm` is awesome golang project !

**Review by marckhouzam — APPROVED** [review]
LGTM Thanks @tanguofu !

This will need approval from another maintainer.
/cc @hickeyma @bacongobbler

**Review by hickeyma — APPROVED** [review]
LGTM, thanks @tanguofu. 

Tested manually using https://github.com/thomastaylor312/advanced-helm-demos/tree/master/post-render where first tested without passing the args (as per the script) and then using the args instead of the script.

**Pull request merged**
